From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Wed, 11 Mar 2026 13:28:41 +0100
Subject: printdialog: Keep GTask alive for portal repsonse

We might return/unref the GTask before the portal response.
The portal response handler uses the GTask data so it must be kept alive.

Fixes: a78873ae56 ("Plug GTask leaks")

Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/9631
---
 gtk/gtkprintdialog.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/gtk/gtkprintdialog.c b/gtk/gtkprintdialog.c
index 94494d6..c51486b 100644
--- a/gtk/gtkprintdialog.c
+++ b/gtk/gtkprintdialog.c
@@ -1160,8 +1160,6 @@ print_response (GDBusConnection *connection,
           break;
         }
     }
-
-  g_object_unref (task);
 }
 
 static void
@@ -1204,7 +1202,8 @@ print_called (GObject      *source,
                                             NULL,
                                             G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
                                             print_response,
-                                            task, NULL);
+                                            g_object_ref (task),
+                                            g_object_unref);
 
     }
 
@@ -1218,8 +1217,8 @@ print_called (GObject      *source,
       ptd->has_returned = TRUE;
       g_object_add_weak_pointer (G_OBJECT (ptd->stream), (gpointer *)&ptd->stream);
       g_task_return_pointer (task, ptd->stream, g_object_unref);
-      g_object_unref (task);
     }
+  g_object_unref (task);
 }
 
 static void
@@ -1255,7 +1254,8 @@ print_window_handle_exported (GtkWindow  *window,
                                         NULL,
                                         G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
                                         print_response,
-                                        task, NULL);
+                                        g_object_ref (task),
+                                        g_object_unref);
 
   fd_list = g_unix_fd_list_new ();
   idx = g_unix_fd_list_append (fd_list, ptd->fds[0], NULL);
